### PR TITLE
Add Sessions and AuthenticateWithPassword to UserManagement

### PR DESCRIPTION
**Pull request by pliao28**
## Description
Add Sessions and related objects to UserManagement
Add AuthenticateWithPassword

## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[ ] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

File: src/WorkOS.net/Services/UserManagement/Entities/Reasons.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        /// The type of reason for unauthorized access to an organization.
+        /// </summary>
+        [JsonProperty("type")]
+        public Type Type { get; set; }
+
+        /// <summary>
+        /// Array of allowed authentication methods allowed by the organziation.
+        /// </summary>
+        [JsonProperty("allowed_authentication_methods")]
+        public AuthenticationMethods[] AllowedAuthenticationMethods { get; set; }
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Entities/Reasons.cs:22]: Using a `List` here would be more idiomatic:

```suggestion
        public List<AuthenticationMethods> AllowedAuthenticationMethods { get; set; }
```

File: src/WorkOS.net/Services/UserManagement/Enums/Type.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+{
+    using System.Runtime.Serialization;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    /// <summary>
+    /// An enum describing the type of reason for unauthorized access to an organization. <see cref="User"/>.
+    /// </summary>
+    [JsonConverter(typeof(StringEnumConverter))]
+    public enum Type
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Enums/Type.cs:11]: Could we give this a more descriptive name?

```suggestion
    public enum SessionUnauthorizedOrganizationReason
```

Comment (pliao28) [reply src/WorkOS.net/Services/UserManagement/Enums/Type.cs:11]: Would we want to rename this field from `type` across the board then?
CleanShot 2023-08-18 at 12 15 21@2x

Comment (maxdeviant) [reply src/WorkOS.net/Services/UserManagement/Enums/Type.cs]: No, the field name should still be `Type`, it is just the enum name that should be changed.

File: src/WorkOS.net/Services/UserManagement/Enums/AuthenticationMethods.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{
+    using System.Runtime.Serialization;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    /// <summary>
+    /// An enum describing the state of a <see cref="User"/>.
+    /// </summary>
+    [JsonConverter(typeof(StringEnumConverter))]
+    public enum AuthenticationMethods
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Enums/AuthenticationMethods.cs:11]: Enums should be named in the singular, as they can only have one inhabitant value at a time:

```suggestion
    public enum AuthenticationMethod
```

File: src/WorkOS.net/Services/UserManagement/Enums/AuthenticationMethods.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+namespace WorkOS
+{
+    using System.Runtime.Serialization;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    /// <summary>
+    /// An enum describing the state of a <see cref="User"/>.
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Enums/AuthenticationMethods.cs:8]: Is this comment accurate?

File: src/WorkOS.net/Services/UserManagement/Entities/Session.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        /// The token which can be used to verify the user's session.
+        /// </summary>
+        [JsonProperty("token")]
+        public string Token { get; set; }
+
+        /// <summary>
+        /// List of organizations with authorized access that the user is a member of.
+        /// </summary>
+        [JsonProperty("authorized_organizaitons")]
+        public AuthorizedOrganizations[] AuthorizedOrganizations { get; set; }
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Entities/Session.cs:40]: Can we use `List` here as well?

```suggestion
        public List<AuthorizedOrganizations> AuthorizedOrganizations { get; set; }
```

In .NET we almost always want to use `List` over `Array`.

File: src/WorkOS.net/Services/UserManagement/Entities/AuthorizedOrganizations.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+namespace WorkOS
+{
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Organizations user has authorized access to.
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Entities/AuthorizedOrganizations.cs:6]:
```suggestion
    /// An organization the user has authorized access to.
```

File: src/WorkOS.net/Services/UserManagement/Entities/UnauthorizedOrganizations.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        /// Organization user does not have authorization to.
+        /// </summary>
+        [JsonProperty("organization")]
+        public Organization Organization { get; set; }
+
+        /// <summary>
+        /// Reason user does not have authorization to organization.
+        /// </summary>
+        [JsonProperty("reasons")]
+        public Reasons[] Reasons { get; set; }
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Entities/UnauthorizedOrganizations.cs:22]:
```suggestion
        public List<Reason> Reasons { get; set; }
```

File: src/WorkOS.net/Services/UserManagement/Entities/Reasons.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+namespace WorkOS
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Reasons for unauthorized access to an organization the user is a member of.
+    /// </summary>
+    public class Reasons
```
Comment (maxdeviant) [inline src/WorkOS.net/Services/UserManagement/Entities/Reasons.cs:10]: Seems like this should be named in the singular?

```suggestion
    public class Reason
```

Comment (pliao28) [reply src/WorkOS.net/Services/UserManagement/Entities/Reasons.cs]: Agree, I was just following doc naming but realized the entities can remain singular 
CleanShot 2023-08-21 at 12 02 33@2x

**Comment by pliao28** [comment]
@maxdeviant  Shall I change the organization domain from array to list in the organization object to follow same practices?
CleanShot 2023-08-21 at 12 05 32@2x

**Comment by maxdeviant** [comment]
> @maxdeviant Shall I change the organization domain from array to list in the organization object to follow same practices? CleanShot 2023-08-21 at 12 05 32@2x

Yea, if we're releasing a new major version we can update those spots as well. Let's keep that in a separate PR.

**Review by maxdeviant — APPROVED** [review]

**Pull request merged**